### PR TITLE
feat: introduce `bptTopup` method for siphoning eure into bpt

### PR DESCRIPTION
**Pull request by petrovska-petro**
todo:

- [x] define logic to determine "excess" in `checker`
- [x] ensure that multicall works as expected in this case (batching: approval & join)
- [x] test whole flow of `bptTopup` method

### CODE REVIEW

File: test/BaseFixture.sol [hunk 1]
```diff
@@ -23,6 +23,7 @@ contract BaseFixture is Test {
 
     uint256 constant EURE_TO_MINT = 1_000e18;
     uint128 constant MIN_EURE_ALLOWANCE = 200e18;
+    uint256 constant EURE_BUFFER = 50e18;
```
Comment (gosuto-inzasheru) [inline test/BaseFixture.sol:26]: im thinking it makes sense to simply make the buffer equal to the daily limit. that way technically there wont be a need for any action more than once a day

**Review by gosuto-inzasheru — COMMENTED** [review]
getting two fails:
```
Ran 4 test suites in 11.33s (22.06s CPU time): 4 tests passed, 2 failed, 0 skipped (6 total tests)

Failing tests:
Encountered 1 failing test in test/TopupEUReTest.t.sol:TopupTest
[FAIL. Reason: assertion failed] testTopupChecker() (gas: 55667)

Encountered 1 failing test in test/TopupTest.t.sol:TopupTest
[FAIL. Reason: assertion failed] testTopupChecker() (gas: 55667)
```

both times it seems to occur on the `assertFalse(canExec);`

also im thinking of doing a var/comment renaming after this merge.

i think it will help readability to define some terms in an intro comment and use them consistently throughouth the contract. rn im thinking:
- safe daily spend (as set in gp ui)
- safe buffer or float (additional funds above daily spend that wont get deposited into the pool)
- safe surplus (funds above buffer/float)
- safe deficit (funds missing to reach safe daily spend)
- the action of siphoning/depositing/topping up/etc needs uniform language

File: test/BaseFixture.sol (hunk 1 above)
Comment (gosuto-inzasheru) [reply test/BaseFixture.sol:26]: wdyt? so lower bound = daily limit, upper bound = daily limit * 2

**Comment by petrovska-petro** [comment]
interesting, i only run `forge test --mt testTopupBpt` locally before pushing as was my focus did not think will break the other suits, but makes sense as there are new conditions to make it `true`. will push fixes today

Screenshot 2024-05-08 at 14 29 36

File: test/BaseFixture.sol (hunk 1 above)
Comment (petrovska-petro) [reply test/BaseFixture.sol:26]: i will prefer to allow giving user flexibility on determining what they think is appropriate for them as "buffer", but the simply daily x 2 can also work as well for poc

**Comment by petrovska-petro** [comment]
now should be g2g 

Screenshot 2024-05-08 at 17 58 14

**Review by gosuto-inzasheru — APPROVED** [review]
passing now

`bytes4(execPayload)` check is a great addition to keep the `assertTrue` valuable

**Pull request merged**